### DISCUSSION TITLE
Garantir certaines assertions sur les règles

### DISCUSSION THREAD

**Discussion by lajarre** [publicodes/publicodes#128] [Ideas]
Aujourd'hui, les règles représentent un graphe de fonctions entre les variables.
On essaie autant que possible de s'assurer qu'il n'y a pas de cycle dans ce
graphe. Supposons pour l'instant qu'il n'y a pas de cycle (pour simplifier la
discussion).

Ainsi on a un DAG, qui a donc certains noeuds qui sont des entrées. Ce sont
notamment les "pivots", comme par exemple le `revenu professionnel`, qui sont
des noeuds d'entrée à forte adjacence.

Il est facile de parcourir le DAG dans le sens descendant, en partant de
variables définies et en construisant petit-à-petit des variables plus "bas" dans le graphe grâce à leur formule. On peut representer cela par `y = f(x1, x2)` avec x1 et x2 déjà définis à l'itération n du parcours du graphe et y défini à l'itération n + 1.

Le problème est que, dans la vraie vie (dans les textes), on écrit les règles
dans le sens du "bon sens" (ou disons de la logique des textes), qui coïncide
rarement avec le sens du DAG. On retrouve cela par exemple dans la définition du `dirigeant . rémunération totale` défini tel que `= indépendant . revenu net de cotisations + indépendant . cotisations et contributions`, alors que le bon sens voudrait écrire comme dans un résultat comptable: `revenu net de cotisations = rémunération totale - cotisations et contributions`.

Ceci pose deux problèmes:

- il est parfois très difficile de lire les règles et d'en faire sens
- lors du développement d'un nouvel ensemble de règles, il est difficile de
    vérifier qu'on a bien suivi les textes.

On peut penser à plusieurs approches:

- créer une API au moteur du type: show("revenu net de cotisations") qui rendrait
    "rémunération totale - cotisations et contributions", autrement dit qui
    inverserait les formules pour présenter la variables définie à partir des
    autres
- permettre au développeur de rédiger des assertions et créer une API au moteur
    qui vérifierait ces assertions. Une telle assertion serait par exemple
    "revenu net de cotisations == rémunération totale - cotisations et contributions".

La 2e approche semble la plus praticable. On voit 2 tactiques:

- soit faire du batch testing (àla Haskell QuickCheck)
- soit utiliser un outil de preuve comme z3.

Exemple avec z3 (python):

```python
rnc = Real("revenu net de cotisations")
rt = Real("remuneration totale")
cc = Real("cotisations et contributions")
traduction_texte_de_loi = rnc == rt - cc
regle_publicode_ok = rt == rnc + cc
regle_publicode_fail = rt == rnc - cc
```

`prove(traduction_texte_de_loi == regle_publicode_ok)` donne "proved"
`prove(traduction_texte_de_loi == regle_publicode_fail)` donne:

```
counterexample
[cotisations et contributions = -1/4,
 revenu net de cotisations = -1/4,
 remuneration totale = 0]
```

Cela depend de pas mal de choses:

- on a un soucis avec l'inversion numerique qui fait peter le DAG, a creuser
    dans le cadre de l'issue des cycles
- on voudrait que les "formules" des regles utilisent des operateurs simples
    qu'on puisse mettre dans z3. Cela suppose de simplifier le langage.

A creuser, mais il me semble qu'il faudrait que les known-mecanisms soient réécrits en publicode et non pas en js directement.

**Comment by lajarre** [comment]
@johangirod @mquandalle et peut-etre @laem je vous laisse jeter un oeil.

Je vois que vous vous etes deja posé des questions similaires dans https://github.com/betagouv/mon-entreprise/issues/718 notamment

**Comment by mquandalle** [comment]
Dans le cas spécifique des règles indépendants, peut-être les règles sont elles effectivement définies "à l'envers" par rapport aux définitions usuelles, et que l'on devrait écrire `net = total - cotisations`, plutôt que `total = net + cotisations`. C'est ce que l'on fait côté salarié, et ça ne doit pas poser de problème au moteur : 

https://github.com/betagouv/mon-entreprise/blob/7cf87ea511cd484a462e965f86d2901eccdcd669/source/rules/salari%C3%A9.yaml#L1583-L1589

**Comment by johangirod** [comment]
Pour moi, les deux approches ont leur pertinences et sont complémentaire. 

Je pense effectivement comme @mquandalle qu'il faudrait parvenir à écrire les règles de la manière la plus proche des définitions usuelle, quitte à devoir améliorer le moteur si on rencontre des limites. C'est la solution actionnable à court terme.

Ceci étant dit, j'aime également beaucoup l'idée d'avoir des invariants que l'on puisse prouver sur notre base de règles. Cela rejoint la discussion entamée dans betagouv/publicodes#6, et requiert un gros investissement pour asseoir le langage dans un système formel permettant l'élaboration de preuve. 

Cela passe notamment par l'utilisation de brique de base plus simples pour la définition des mécanismes (nous avons déjà évoqué l'idée avec @mquandalle, mais je ne parviens pas à retrouver la trace de cette discussion).

Cette redéfinition pourrait permettre une inversion analytique, et donc d'avoir une formule associée pour les deux directions de chaque arrête du graphe (quand c'est mathématiquement possible).

Ce n'est pas une mince affaire, mais c'est un cap à garder en tête si on veut faire de ce langage la référence pour l'implémentation des algorithmes d'intérêt publique.

**Comment by lajarre** [comment]
> C'est ce que l'on fait côté salarié, et ça ne doit pas poser de problème au moteur :

Ok je vois, mais le fait est que la partie independants etait tres difficile a lire. Est-ce que ca meriterait donc une refacto pour etre "dans le sens de la loi"? Je garde ca dans un coin de mon esprit le temps de maturer tout ca.

Ceci dit je crains que d'une maniere generale, il n'y ait pas un *sens de la loi*, mais **des** *sens de la loi*. Cf. ma reflexion sur les synonymes dans betagouv/mon-entreprise#964 (opposition des regles comptables aux regles fiscales, portant pourtant sur les memes variables). Ainsi, l'idee de verifier une assertion sur les regles (verifier qu'une equation est prouvable) est toujours valable. Mais ici aussi, j'attends de valider le besoin en gagnant plus d'experience operationnelle.

**Comment by lajarre** [comment]
> Cela passe notamment par l'utilisation de brique de base plus simples pour la définition des mécanismes (nous avons déjà évoqué l'idée avec @mquandalle, mais je ne parviens pas à retrouver la trace de cette discussion).

Oui c'est clairement par la qu'il faut commencer. Si tu retrouves la discussion a l'occasion (rien ne presse), je suis interesse.

> Cette redéfinition pourrait permettre une inversion analytique, et donc d'avoir une formule associée pour les deux directions de chaque arrête du graphe (quand c'est mathématiquement possible).

Tu vas un cran plus loin que moi ici, et c'est peut-etre d'une difficulté plus grande. Dans mon cas, je veux simplement verifier qu'un systeme d'equations sur des nombres reels sont prouvables ou non (admettent des valeurs qui rendent _vrai_ ou non). Je n'ai pas demande a Z3 de me pondre l'equation inversee, mais seulement de verifier un systeme d'equations dans lequel j'ai moi-meme, developpeur publicode, inscrit une equation supplementaire. Mais j'arrive aux limites de mes connaissances actuelles des outils de preuve, peut-etre qu'on pourrait leur demander de faire l'inversion eux-memes, a creuser...

> Ce n'est pas une mince affaire, mais c'est un cap à garder en tête si on veut faire de ce langage la référence pour l'implémentation des algorithmes d'intérêt publique.

En effet j'ai aussi cette intuition. J'ai l'impression qu'une bonne premiere etape serait de reecrire les known-mecanisms en publicode, non?

**Comment by denismerigoux** [comment]
Bonjour, betagouv/publicodes#6 a été tagguée donc je viens et je trouve la discussion très intéressante ! J'en profite pour sortir une citation : 
> Il est parfois très difficile de lire les règles et d'en faire sens

> Lors du développement d'un nouvel ensemble de règles, il est difficile de vérifier qu'on a bien suivi les textes.

C'est tout à fait le diagnostic que j'avais effectué pour des langages _à la M_ comme Publicodes tel qu'il est actuellement. C'est pour cette raison que je me suis tourné vers le paradigme de *programmation littéraire* et l'annotation directe des textes de loi avec le code. Voici à quoi ça ressemble dans notre travail en cours (exemple des allocations familiales) : 

```
@Article L512-3@
Sous réserve des règles particulières à chaque prestation, ouvre droit aux prestations familiales :

1°) tout enfant jusqu’à la fin de l’obligation scolaire ;
/*
champ d'application PrestationsFamiliales :
  règle sous condition
    existe enfant dans ménage.enfants tel que
      date_aujourd_hui <= enfant.fin_obligation_scolaire
  conséquence l512_3.droits_ouverts_prestations_familiales rempli
*/

2°) après la fin de l’obligation scolaire, et jusqu’à un âge limite, tout enfant dont la rémunération éventuelle n’excède pas un plafond
/*
champ d'application PrestationsFamiliales :
  règle sous condition
    existe enfant dans ménage.enfants tel que
      date_aujourd_hui > enfant.fin_obligation_scolaire et
      enfant.rémunération <= l512_3.plafond_rémunération_alinéa_2 et
      enfant.âge < l512_3.âge_limite_alinéa_2
  conséquence l512_3.droits_ouverts_prestations_familiales rempli
*/
```

Votre idée d'assertion me semble bonne, je l'avais intégrée dès le début dans ma réflexion sur le langage. Je m'en sers aussi pour des trucs comme ça :

```
Le montant des allocations mentionnées aux deux premiers alinéas du présent article, ainsi que celui des majorations mentionnées à l'article L. 521-3 varient en fonction des ressources du ménage ou de la personne qui a la charge des enfants, selon un barème défini par décret.
/*
champ d'application AllocationsFamiliales :
  assertion l512_1.allocations_familiales fixé par décret
  assertion l512_1.allocations_familiales varie avec ménage.ressources
  assertion l521_3.majorations_allocations_familiales fixé par décret
  assertion l512_1.majorations_allocations_familiales varie avec
     ménage.ressources
*/
```

Concernant la puissance des outils de preuve actuels, vous avez raison : inverser analytiquement une formule d'arithmétique non-linéaire est un problème assez intractable ; Z3 va vous donner une solution au cas par cas mais ça risque de prendre du temps et il vous faudra sûrement faire des hypothèses supplémentaires comme borner tous vos entiers par une valeur maximale (pour encoder sur des vecteurs de bits). J'ai moi-même fait exactement ça dans mon étude du taux marginal effectif (http://gallium.inria.fr/seminaires/transparents/20191104.Denis.Merigoux.pdf).

Cependant cette histoire d'inversion me semble cruciale, et j'ai l'intention de m'y attaquer également. Mais pour ça je compte utiliser tout l'arsenal des méthodes formelles : optimisations classiques de compilation, interprétation abstraite, et en dernier recours génération d'obligations de preuve Z3 ou même Coq.

En attendant cela, simplement écrire les assertions dans le code et les vérifier lors de l'exécution est déjà un bon premier pas :)

**Comment by denismerigoux** [comment]
Et vous avez raison pour la réécriture de vos mécanismes à partir de briques de base "simples", c'est exactement ça le travail de formalisation du langage que j'essaie de vous décrire dans betagouv/publicodes#6.

**Comment by laem** [comment]
> On retrouve cela par exemple dans la définition du dirigeant . rémunération totale défini tel que = indépendant . revenu net de cotisations + indépendant . cotisations et contributions, alors que le bon sens voudrait écrire comme dans un résultat comptable: revenu net de cotisations = rémunération totale - cotisations   et contributions.

À ma connaissance le calcul du revenu et cotisations des indépendants est complètement tordu et repose par définition sur une inversion. Étant donné que cotisations = barème(revenu net), le "sens" d'écriture `revenu net de cotisations = rémunération totale - cotisations` n'est pas forcément le plus intuitif. 

Ce n'est pas le cas pour le calcul salarié, beaucoup plus simple, d'où le fait qu'on l'ait écrit dans l'autre sens.

Mais ça n'enlève pas comme tu le dis @lajarre le besoin éventuel de vérifier des assertions.
